### PR TITLE
Data: Optimize withSelect to avoid generating merge props on equal props

### LINKED ISSUE #6466: Slow perceived performance compared to Classic Editor
First, try typing a couple paragraphs in Gutenberg. Then, try typing the same paragraphs in Classic Editor.

When you experience these two experiences side by side, Gutenberg is noticeably slower than the Classic Editor. In fact, for any amount of substantial writing, the Classic Editor is refreshing to use over Gutenberg.

I'd suggest we:

* [ ] Find some way of benchmarking this experience as a performance metric.
* [ ] Track down and address the source of slowness. There could be a silver bullet, or it could be a variety of factors.
* [ ] Ideally, work the performance metric into our automated testing system.

### LINKED ISSUE #7498: High CPU usage when simply typing in paragraph block
**Describe the bug**
Ever since the latest update to 3.1.0, typing in the Gutenberg editor thrashes my CPU.  I've tested this on my local site and also using [Frontenberg](https://frontenberg.tomjn.com/) (just to see if it was my setup).  Whenever I stop typing, CPU usage drops dramatically.  I don't remember this being an issue with the previous version, but it at least wasn't as bad because I didn't notice.  Right now I can't even type without serious lag.

**To Reproduce**
1. Open a new post or edit an old post
2. Type in a Gutenberg text box
3. Watch CPU usage

**Expected behavior**
I would expect just typing to not use 100% of my CPU and lag so much that I can't see what I'm typing without buffering.

**Screenshots**
gutenburnthecpu

**Desktop (please complete the following information):**
 - OS: Microsoft Windows [Version 10.0.16299.492]
 - Browser: Google Chrome
 - Version: 67.0.3396.87 (Official Build) (64-bit)

### LINKED ISSUE #7680: Typing becomes more laggy the more blocks you have in a post
**Describe the bug**

Typing can become more laggy, and possibly totally unresponsive, the more blocks you have in a post.

**To Reproduce**
1. Start a new post with a single paragraph block.
2. Hold down a key. The key should repeat in the paragraph block and the browser should stay responsive.
3. Create an extra 30 paragraph blocks.
4. Hold down a key in one of the paragraph blocks. The editor will lag, and it's possible for the browser to freeze.
5. Take your finger off the key, and notice how long it takes for the editor to update.

**Expected behavior**

Typing in a paragraph block should be responsive no matter how many blocks you have in a post.

**Screenshots**

lag

**Desktop (please complete the following information):**
 - Browser: Chrome, Firefox, Safari
 - Version: all latest versions

**Smartphone (please complete the following information):**
 - Device: Hauwei mate 10 pro - noticeable lag even with just one block

### PR DESCRIPTION
**Pull request by aduth**
Closes #7680 
Closes #6466 
Closes #7498

This pull request seeks to implement the data module's `withSelect` implementation to avoid a case where `mapSelectToProps` would be called wastefully when neither the state had triggered a change nor had incoming props changed (as a product of shallow equality comparison).

This had been effected in previous implementations leveraging `componentWillReceiveProps`, but was lost in the upgrade to using `static getDerivedStateFromProps`:

https://github.com/WordPress/gutenberg/blob/95ef796aa5f067c7c785573295f3a56a20648e9f/packages/data/src/index.js#L300-L301

It required some refactoring to account for state changes made by the artificial trigger from the component's subscription, to handle generating new merge props on either a state change or incoming props change, while ensuring that the component only re-renders when either props or merge props are shallowly unequal, also ensuring that `isShallowEqual` is called at most one time per props sets during this render pass.

In future iterations, it may be worth exploring to see whether `props` and `fromStoreUpdate` can be omitted from state. I'd considered assigning instance variables in `shouldComponentUpdate` and the subscription callback, where in the latter we'd call `forceUpdate` to skip `shouldComponentUpdate`. The main downside to this is that `shouldComponentUpdate` should ideally not have side effects such as assigning instance variables.

**Testing instructions:**

Ensure new unit tests pass:

```
npm run test-unit
```

Observe lag or CPU usage differences from this branch in contrast with master.

### CODE REVIEW

**Comment by aduth** [comment]
While this is likely a valid optimization, in some testing this is still not even close to the same level of performance. Typing in the demo post, for example, is unbearably slow, even after these changes, when contrasted with the [variation of `withSelect` from 95ef796](https://github.com/WordPress/gutenberg/blob/95ef796aa5f067c7c785573295f3a56a20648e9f/packages/data/src/index.js#L269-L344).

**Comment by aduth** [comment]
>In future iterations, it may be worth exploring to see whether `props` and `fromStoreUpdate` can be omitted from state. I'd considered assigning instance variables in `shouldComponentUpdate` and the subscription callback, where in the latter we'd call `forceUpdate` to skip `shouldComponentUpdate`. The main downside to this is that `shouldComponentUpdate` should ideally not have side effects such as assigning instance variables.

This approach turned out to be much more viable, and made a night-and-day difference to the usability of the demo post. I'm still not entirely clear what was problematic about the previous implementation: It may be that `getDerivedStateFromProps` is simply not performant, or due to the additional memory overhead of tracking `props` and `fromStoreChange` in state.

**Review by mcsf — APPROVED** [review]
The last iteration of this PR makes a huge difference for me when editing the demo post.

I'd have a second set of eyes on this, but it looks like solid work. Yes, concessions are made, but this is quite the hot path.

File: packages/data/src/index.js [hunk 1]
```diff
@@ -266,27 +266,36 @@ export function dispatch( reducerKey ) {
...
+	/**
+	 * Given a props object, returns the next merge props by mapStateToProps.
+	 *
+	 * @param {Object} props Props to pass as argument to mapStateToProps.
+	 *
+	 * @return {Object} Props to merge into rendered wrapped element.
+	 */
+	function getNextMergeProps( props ) {
+		return (
+			mapStateToProps( select, props ) ||
```
Comment (gziolo) [inline packages/data/src/index.js]: Unrelated: we should rename `mapStateToProps` to `mapSelectToProps`.

**Comment by gziolo** [comment]
Yes, I can confirm that it works much better. I was able to confirm that we don't notice double re-renders anymore in both testing and when comparing with the `master` branch.
Although proposed code uses some not-recommended solution as noted in the inline comment, it is much easier to follow and in my opinion works exactly as we want.

**Pull request merged**